modules/azure/nsg.go

```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		return network.SecurityRulesClient{}, err
+	}
+
+	nsgClient := network.NewSecurityRulesClient(subscriptionID)
+
+	// Get an authorizer
+	auth, err := NewAuthorizer()
+	if err != nil {
+		return network.SecurityRulesClient{}, err
+	}
+
+	nsgClient.Authorizer = *auth
+	return nsgClient, nil
+}
+
+// GetAllNSGRulesE returns a slice containing all rules from the NSG (including defauls and custom)
+func GetAllNSGRulesE(resourceGroupName, nsgName, subscriptionID string) (NsgRuleSummaryList, error) {
+	defaultRulesClient, err := GetDefaultNsgRulesClientE(subscriptionID)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	customRulesClient, err := GetCustomNsgRulesClientE(subscriptionID)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	defaultRuleList, err := defaultRulesClient.ListComplete(context.Background(), resourceGroupName, nsgName)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	customRuleList, err := customRulesClient.ListComplete(context.Background(), resourceGroupName, nsgName)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	rules := make([]NsgRuleSummary, 0)
+	bindRuleList(&rules, defaultRuleList)
+	bindRuleList(&rules, customRuleList)
+	ruleList := NsgRuleSummaryList{}
+	ruleList.SummarizedRules = rules
+	return ruleList, nil
+}
+
+// NsgRuleSummaryList holds a colleciton of NsgRuleSummary structs
+type NsgRuleSummaryList struct {
+	SummarizedRules []NsgRuleSummary
+}
+
+// NsgRuleSummary is a string-based summary of an NSG rule with methods attached.
+type NsgRuleSummary struct {
+	Name                     string
+	Description              string
+	Protocol                 string
+	SourcePortRange          string
+	DestinationPortRange     string
+	SourceAddressPrefix      string
+	DestinationAddressPrefix string
+	Access                   string
+	Priority                 int32
+	Direction                string
+}
+
+func bindRuleList(dest *[]NsgRuleSummary, source network.SecurityRuleListResultIterator) {
+	for source.NotDone() {
+		v := source.Value()
+		*dest = append(*dest, convertToNsgRuleSummary(v.Name, v.SecurityRulePropertiesFormat))
+		err := source.NextWithContext(context.Background())
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func convertToNsgRuleSummary(name *string, rule *network.SecurityRulePropertiesFormat) NsgRuleSummary {
+	summary := NsgRuleSummary{}
+
+	if rule.Description != nil {
+		summary.Description = *rule.Description
+	} else {
+		summary.Description = ""
+	}
+
+	summary.Name = *name
+	summary.Protocol = string(rule.Protocol)
+	summary.SourcePortRange = *rule.SourcePortRange
+	summary.DestinationPortRange = *rule.DestinationPortRange
+	summary.SourceAddressPrefix = *rule.SourceAddressPrefix
+	summary.DestinationAddressPrefix = *rule.DestinationAddressPrefix
+	summary.Access = string(rule.Access)
+	summary.Priority = *rule.Priority
+	summary.Direction = string(rule.Direction)
```

Comment:
I'm a bit worried about all these pointer dereferences without `nil` checks leading to a `panic`... The AWS SDK has a bunch of methods that safely dereference these sorts of things: e.g., `aws.StringValue(XXX)` takes a string pointer `XXX` and, if it's not `nil`, dereferences it, and if it is `nil`, returns an empty string. Perhaps do something similar here?

Reply:
Added safe helpers to assist with handling the case of nil pointer references.